proposals/NativeInt.md

```diff
@@ -0,0 +1,263 @@
+
+Compound assignment operations `x op= y` where `x` or `y` are native ints follow the same rules as with other primitive types with pre-defined operators.
+Specifically the expression is bound as `x = (T)(x op y)` where `T` is the type of `x` and where `x` is only evaluated once.
+
+### dynamic
+
+The conversions and operators are synthesized by the compiler and are not part of the underlying `IntPtr` and `UIntPtr` types.
+As a result those conversions and operators _are not available_ from the runtime binder for `dynamic`. 
+
+```C#
+nint x = 2;
+nint y = x + x; // ok
+dynamic d = x;
+nint z = d + x; // RuntimeBinderException: '+' cannot be applied 'System.IntPtr' and 'System.IntPtr'
+```
+
+### Type members
+
+The only constructor for `nint` or `nuint` is the parameter-less constructor.
+
+The members of `System.IntPtr` and `System.UIntPtr` other than the constructors and operators are available from `nint` and `nuint`.
+For `nint` the following `System.IntPtr` members are available. The members available on `nuint` are similar.
+
+```C#
+public static readonly IntPtr Zero;
+public static int Size { get; }
+public static IntPtr Add(IntPtr pointer, int offset);
+public static IntPtr Subtract(IntPtr pointer, int offset);
+public override bool Equals(object obj);
+public override int GetHashCode();
+public int ToInt32();
+public long ToInt64();
+public void* ToPointer();
+public override string ToString();
+public string ToString(string format);
+```
+
+Some members above use `IntPtr` in the signature even when used from `nint`.
+
+```C#
+nint x = nint.Zero;      // nint x
+nint y = nint.Add(x, 1); // nint y
+var z = nint.Zero;       // IntPtr z
+var w = nint.Add(x, 1);  // IntPtr w
+```
+
+Interfaces implemented by `System.IntPtr` and `System.UIntPtr` are available from `nint` and `nuint`.
+
+```C#
+nint n = 42;
+IEquatable<nint> i = n; // ok, IntPtr implements IEquatable<IntPtr>
+```
+
+### Overriding, hiding, and implementing
+
+The `native int` aliases and the underlying types are considered equivalent for overriding, hiding and implementing.
+
+Overloads cannot differ by `native int` and underlying type alone.
+Overrides and implementations may differ by `native int` and underlying type alone.
+Methods hide other methods that differ by `native int` and underlying type alone.
+
+### Miscellaneous
+
+`native int` types cannot be used as enum underlying types.
+```C#
+enum E : nint // error
+{
+}
+```
+
+`default(nint)` and `new nint()` are equivalent to `(nint)0`.
+
+`typeof(nint)` is `typeof(IntPtr)`.
+
+`sizeof(nint)` is supported and does not require compiling in an unsafe context. The value is not a compile-time constant. `sizeof(nint)` is implemented as `sizeof(IntPtr)` rather than `IntPtr.Size`.
+
+Compiler diagnostics for type references involving `nint` or `nuint` report `nint` or `nuint` rather than `IntPtr` or `UIntPtr`.
+
+### Metadata
+
+`nint` and `nuint` are represented in metadata as `System.IntPtr` and `System.UIntPtr`.
```

Comment:
Do I understand correctly that there is no longer any reason to use `IntPtr` and `UIntPtr`? Even public API can be changed, it seems. We could search and replace our entire codebase?

Reply:
There may be some reasoning to continue using `IntPtr` if users want to clearly differentiate something that is meant to be an "opaque handle" vs something that is meant to be a "native integer". Something that is meant to be a "pointer" could fall into either category, depending on what operations you need to perform on it.

Reply:
An opaque handle is often not a pointer, even under the covers. So the confusion still exists (possibly in my mind only)

Reply:
Right, opaque handles are often something you never want to perform arithmetic with; and so they are good candidates to stay as `IntPtr`.

Things which are actually `pointers` you may want to make `nint`, if performing arithmetic is something needed.

Things which are actually `numbers` you probably want to make `nint`, since you want to treat them as numbers.

Reply:
Right, hence we're continuing to recommend something that contains `Ptr` in the name for representing things that are not necessarily pointers. That's a shame, but perhaps it isn't a problem this set out to solve.